README.md
# Jets-of-Time-Tracker

A [Chrono Trigger: Jets of Time](https://ctjot.com) tracker pack for [PopTracker](https://github.com/black-sliver/PopTracker), with autotracking.

## Usage

Download a .zip (or .tar.gz) from [the Releases page](
https://github.com/coffeemancy/Jets-of-Time-Tracker/releases).

Extract this into your PopTracker packs directory per [PopTracker "Getting started"](
https://github.com/black-sliver/PopTracker#getting-started).

### Variants

This tracker pack includes several variants:

* Item Tracker: item-only tracker for Standard mode
* Legacy of Cyrus: item and map tracker for Legacy of Cyrus mode
* Item Tracker (Lost Worlds): item-only tracker for Lost Worlds mode
* Map Tracker (Lost Worlds): item and map tracker for Lost Worlds mode
* Map Tracker: item and map tracker for Standard mode
* Vanilla Rando: item and map tracker for Vanilla Rando mode

Map Tracker variants allow for setting additional flags, such as Chronosanity
and Epoch Fail. They also support "Broadcast View" (opens another smaller window
with just the items).

### Autotracking

This pack is compatible with [PopTracker SNES game autotracking](https://github.com/black-sliver/PopTracker#auto-tracking).


changelog.txt
v2.2.0
  - Add support for Ice Age mode
  - Add support for vertical tracker orientation
  - Add support for extra spots: Bekkler, Cyrus Grave, Ozzie's Fort, Race Log
  - Add support for extras KIs: Johnny Race (bikekey), Restore Tools (tools)
  - Add support for Split Arris Dome (arrisseed), Sun Keep Spot (separate moon/sunstone)
  - Add support for Unlocked Skyways, Rocksanity, Bucket List

v2.1.1
  - Fixed tracking equipped items to work with Duplicate Characters.

v2.1.0
  - Updated for Jets of Time version 3.2
  - Added Vanilla Rando variant
  - Added Legacy of Cyrus variant
  - Combined the map tracker and chronosanity variants
  - Added Epoch Fail flag to all variants except Lost Worlds
  - Added chronosanity flag to all variants

v2.0.2
  - Bug fix for Ozzie's Fort access in chronosanity mode.

v2.0.1
  - Updated for 3.1.1 logic changes.
  - Autotracker will activate the "GO" button when Frog and both halves of the 
      Masamune are acquired instead of waiting for the Masamune to be reforged.

v2.0.0
  - General upates for version 3.0 of the randomizer.
  - Added map tracker and items only variants for Lost Worlds.
  - Added Chronosanity variant.
  - Autotracker updates for Lost Worlds and Chronosanity variants.
  - Reorganized item layout to group items by go mode.
 
v1.2.1
 - The "GO" button will now autotrack when go mode is achieved.
 - Autotracking bugfixes
   - Items will now be properly untracked due to deaths or resets.
   - Fixed a bug causing Rust Tyrano to not autotrack.
   - Fixed a bug causing hero medal to untrack if Frog was wearing it.
   - Masamune 2 is now tracked correctly.
   - Fixed a bug where Death Peak would not properly track with some party compositions.
   - Fixed a bug causing magic to not track correctly in Lost Worlds mode.
 
v1.2.0
 - Added Autotracking functionality (Developed by Anguirel)

v1.1.1
 - Added Kingdom of Zeal to Dark Ages map, adjusted boss location.

v1.1.0
 - Updated logic to match Jets version 1.5!
 - Added "All Eras" map allowing view of all locations.
 - Added "Characters Locked" flag for logic toggle.
 - New Character Icons.

v1.0.1
 - Changed from character portraits to character menu icons.

v1.0.0:
	- Initial Release
	
- Huge thanks go out to Aeralis for the Wings of Time tracker that was the core basis for this tracker.
- Special thank you to Astharis for asset creation/modification and support.


tests/test_json.py
import json

from pathlib import Path
from urllib.parse import urljoin
from typing import Any, Dict, List

import pytest
import requests

from jsonschema import validate

POPTRACKER_PACK_TYPES = ['items', 'layouts', 'locations', 'manifest', 'maps']
POPTRACKER_PACK_STRICT_TYPES = ['manifest', 'maps']
SCHEMA_CACHE_DIR = Path(Path(__file__).parent.resolve(), '.poptracker_schemas')
STRICT_SCHEMA_CACHE_DIR = Path(SCHEMA_CACHE_DIR, 'strict')
POPTRACKER_REMOTE_PATH = 'https://poptracker.github.io/schema/packs/'
POPTRACKER_REMOTE_STRICT_PATH = urljoin(POPTRACKER_REMOTE_PATH, 'strict/')


@pytest.fixture(scope='session')
def jsonfiles() -> List[Path]:
    return [path for path in Path('.').rglob('*.json') if path.parts[0] not in ['.venv', 'tests']]


@pytest.fixture(scope='session')
def formatted_jsonfiles(jsonfiles) -> List[Path]:
    # exclude older paths and gradually update to new formatting
    # to avoid having large git diffs
    excluded_paths = [
        'items/',
    ]
    return [p for p in jsonfiles if not any(str(p).startswith(ep) for ep in excluded_paths)]


@pytest.fixture(scope='session')
def pack_files() -> Dict[str, List[Path]]:
    files_map: Dict[str, List[Path]] = {}
    all_dirs = [p for p in Path('.').rglob('*') if p.is_dir()]
    for pack_type in POPTRACKER_PACK_TYPES:
        jsonfiles = []
        pack_type_dirs = [p for p in all_dirs if str(p) == pack_type]
        for path in pack_type_dirs:
            jsonfiles.extend([f for f in path.rglob('*.json')])
        files_map[pack_type] = jsonfiles
    return files_map


def get_or_download_schemas(cache_dir: Path, remote_path: str) -> Dict[str, Dict[str, Any]]:
    # local schema file cache
    cached_files = {schema: Path(cache_dir, f'{schema}.json') for schema in POPTRACKER_PACK_TYPES}
    schemas: Dict[str, Dict[str, Any]] = {}
    for schema in POPTRACKER_PACK_TYPES:
        try:
            schemas[schema] = json.load(cached_files[schema].open())
        except IOError:
            url = urljoin(remote_path, f'{schema}.json')
            try:
                remote_schema = requests.get(url).json()
            except Exception as ex:
                err = f'Unable to download poptracker schema ({schema}): {url}'
                raise ValueError(err) from ex
            cached_files[schema].write_text(json.dumps(remote_schema))
            schemas[schema] = remote_schema
    return schemas


@pytest.fixture(scope='session')
def poptracker_schemas() -> Dict[str, Dict[str, Any]]:
    '''PopTracker JSON schemas downloaded from remote.'''
    if not SCHEMA_CACHE_DIR.exists():
        SCHEMA_CACHE_DIR.mkdir()

    cache_dir = SCHEMA_CACHE_DIR
    remote_path = POPTRACKER_REMOTE_PATH
    return get_or_download_schemas(cache_dir, remote_path)


@pytest.fixture(scope='session')
def poptracker_strict_schemas() -> Dict[str, Dict[str, Any]]:
    if not SCHEMA_CACHE_DIR.exists():
        SCHEMA_CACHE_DIR.mkdir()
    if not STRICT_SCHEMA_CACHE_DIR.exists():
        STRICT_SCHEMA_CACHE_DIR.mkdir()

    cache_dir = STRICT_SCHEMA_CACHE_DIR
    remote_path = POPTRACKER_REMOTE_STRICT_PATH
    return get_or_download_schemas(cache_dir, remote_path)


def test_all_jsonfiles_loadable(jsonfiles):
    '''Check all files are loadable as JSON.'''
    for jsonfile in jsonfiles:
        assert json.load(jsonfile.open())


@pytest.mark.parametrize('pack_type', POPTRACKER_PACK_TYPES)
def test_pack_schema_validation(pack_type, poptracker_schemas, pack_files):
    schema = poptracker_schemas[pack_type]
    for jsonfile in pack_files[pack_type]:
        try:
            validate(json.load(jsonfile.open()), schema)
        except Exception as ex:
            err = f'Failed to JSON schema validate file: {jsonfile}'
            raise ValueError(err) from ex


@pytest.mark.parametrize('pack_type', POPTRACKER_PACK_STRICT_TYPES)
def test_pack_strict_schema_validation(pack_type, poptracker_schemas, pack_files):
    schema = poptracker_schemas[pack_type]
    for jsonfile in pack_files[pack_type]:
        try:
            validate(json.load(jsonfile.open()), schema)
        except Exception as ex:
            err = f'Failed to strict JSON schema validate file: {jsonfile}'
            raise ValueError(err) from ex


def test_json_file_style(formatted_jsonfiles):
    '''Check all formatted json files are formatted per json.tool.'''
    for jsonfile in formatted_jsonfiles:
        text = jsonfile.read_text()
        loaded_json = json.loads(text)
        formatted_output = json.dumps(loaded_json, indent=2) + '\n'
        try:
            assert text == formatted_output
        except AssertionError as ex:
            err = (
                f'JSON file {jsonfile} is not formatted correctly.\n'
                f'Fix by formatting with:\n'
                f'  python -m json.tool --indent=2 {jsonfile} {jsonfile}\n\n'
            )
            raise ValueError(err) from ex


tests/json/test_json.py
import json

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Type
from urllib.parse import urljoin, urlsplit

import jsonschema
import pytest
import requests

from referencing import Registry, Resource

JSONSchema = Dict[str, Any]


# FIXTURES ###################################################################


@dataclass
class TrackerClient:
    cache: str
    remote_path: str
    pack_types: List[str]


class PopTracker(TrackerClient):
    cache = 'poptracker'
    remote_path = 'https://poptracker.github.io/schema/packs/'
    pack_types = ['items', 'layouts', 'locations', 'manifest', 'maps']


class PopTrackerStrict(TrackerClient):
    cache = 'poptracker'
    remote_path = 'https://poptracker.github.io/schema/packs/strict/'
    pack_types = ['manifest', 'maps']


class EmoTracker(TrackerClient):
    cache = 'emotracker'
    remote_path = 'https://emotracker.net/developers/schemas/'
    pack_types = ['all', 'items', 'layouts', 'locations']


def get_or_download(cache: Path, uri: str) -> JSONSchema:
    '''Get schema from cache file or download and write.'''
    cached_file = Path(cache / urlsplit(uri).path.lstrip('/')).resolve()
    try:
        contents = json.load(cached_file.open())
    except IOError:
        try:
            contents = requests.get(uri).json()
        except Exception as ex:
            err = f"Unable to download tracker schema: {uri}"
            raise ValueError(err) from ex
        cached_file.parent.mkdir(parents=True, exist_ok=True)
        cached_file.write_text(json.dumps(contents))
    return contents


def get_pack_files(root: Path) -> Dict[str, List[Path]]:
    # get all directories that don't start with '.' and aren't in tests
    all_dirs = [path for path in [p for p in root.rglob('*') if p.is_dir()] if is_pack_file(path, root)]

    # group files in directories based on pack_type
    files_map: Dict[str, List[Path]] = {}
    for pack_type in PopTracker.pack_types:
        jsonfiles: List[Path] = []
        pack_type_dirs = [p.relative_to(root) for p in all_dirs if pack_type in p.relative_to(root).parts]
        for path in pack_type_dirs:
            jsonfiles.extend([f for f in path.glob('*.json') if not f.is_symlink()])
        files_map[pack_type] = jsonfiles
    files_map['manifest'] = [Path('manifest.json')]

    return files_map


def is_pack_file(path: Path, root: Path):
    part = path.relative_to(root).parts[0]
    return not part.startswith('.') and part not in ['tests', 'tools']


@pytest.fixture(scope='session')
def jsonfiles(paths) -> List[Path]:
    return [
        path for path in paths['root'].rglob('*.json') if not path.is_symlink() and is_pack_file(path, paths['root'])
    ]


@pytest.fixture(scope='session')
def pack_files(paths) -> Dict[str, List[Path]]:
    return get_pack_files(paths['root'])


@pytest.fixture(
    scope='session',
    params=[PopTracker, PopTrackerStrict, EmoTracker],
    ids=['PopTracker', 'PopTracker[strict]', 'EmoTracker'],
)
def tracker(request) -> Type[TrackerClient]:
    return request.param


@pytest.fixture(scope='session')
def registry(paths, tracker) -> Registry:
    '''JSONSchema registry using locally cached test files.'''

    def retrieve(uri: str):
        return Resource.opaque(get_or_download(paths[tracker.cache], uri))

    return Registry(retrieve=retrieve)


@pytest.fixture(scope='session')
def schemas(paths, tracker) -> Dict[str, JSONSchema]:
    '''Schemas for pack files using locally cached schema files.'''
    jsonschemas: Dict[str, JSONSchema] = {}
    for schema in tracker.pack_types:
        uri = urljoin(tracker.remote_path, f"{schema}.json")
        jsonschemas[schema] = get_or_download(paths[tracker.cache], uri)
    return jsonschemas


@pytest.fixture(scope='session')
def validators(tracker, registry, schemas) -> Dict[str, jsonschema.validators.Validator]:
    '''JSONSchema validators for pack files using locally cached schema files.'''
    vmap: Dict[str, jsonschema.validators.Validator] = {}
    for pack_type in tracker.pack_types:
        schema = schemas.get('all', schemas[pack_type])
        validator_cls = jsonschema.validators.validator_for(schema)
        vmap[pack_type] = validator_cls(schema, registry=registry)
    return vmap


# TESTS ######################################################################


def test_all_jsonfiles_loadable(jsonfiles):
    '''Check all files are loadable as JSON.'''
    assert jsonfiles, 'Failed to find any JSON files.'

    for jsonfile in jsonfiles:
        assert json.load(jsonfile.open())


def test_expected_pack_files(paths, jsonfiles, pack_files):
    '''Coherence check between jsonfiles and pack_files to assure not missing files in tests.

    This test is to assure that if updates in jsonfiles or pack_files fixtures are made,
    that the expected files for testing are not silently missed. All files in jsonfiles
    should be in pack_files fixture (except for expected exceptions, like settings.json).
    '''
    all_pack_files = {file for files in pack_files.values() for file in files}

    # "settings.json" files are in jsonfiles but not pack_files
    expected_differences = ['settings.json']

    expected_pack_files = set()
    for jsonfile in jsonfiles:
        file = jsonfile.relative_to(paths['root'])
        if file.parts[-1] not in expected_differences:
            expected_pack_files.add(file)

    assert all_pack_files == expected_pack_files


@pytest.mark.parametrize('pack_type', PopTracker.pack_types)
def test_pack_schema_validation(pack_type, tracker, validators, pack_files, print_debug):
    '''Check pack files pass JSONSchema validation.'''
    assert pack_files[pack_type], f'Missing {pack_type} pack files!'

    if pack_type not in tracker.pack_types:
        pytest.skip(f"no schema for {pack_type}")

    validator = validators[pack_type]
    print_debug(f"\nUsing {type(validator).__name__} for '{pack_type}' schema...")

    for jsonfile in pack_files[pack_type]:
        print_debug(f"* Validating {jsonfile}")
        try:
            validator.validate(json.load(jsonfile.open()))
        except Exception as ex:
            err = f'Failed to JSON schema validate file: {jsonfile}'
            raise ValueError(err) from ex


def test_json_file_style(paths, jsonfiles):
    '''Check all formatted json files are formatted per json.tool.'''
    for jsonfile in jsonfiles:
        text = jsonfile.read_text()
        loaded_json = json.loads(text)
        formatted_output = json.dumps(loaded_json, indent=2) + '\n'
        try:
            assert text == formatted_output
        except AssertionError as ex:
            rel_jsonfile = jsonfile.relative_to(paths['root'])
            err = (
                f"JSON file {jsonfile} is not formatted correctly.\n"
                f"Fix by formatting with:\n"
                f"  python -m json.tool --indent=2 {rel_jsonfile} {rel_jsonfile}\n\n"
            )
            raise ValueError(err) from ex


tests/test_locations.py
import json
import typing as t

from pathlib import Path

import pytest

MAP_OFFSETS: t.Dict[str, t.Tuple[int, int]] = {
    # Prehistory | Middle Ages | Future
    # Dark Ages  | Present     | End of Time
    'All Eras': {
        'Prehistory': (0, 0),
        'Dark Ages': (0, 1024),
        'Middle Ages': (1536, 0),
        'Present': (1536, 1024),
        'Future': (3072, 0),
        'End Of Time': (3072, 1024),
    },
    # Prehistory  | Dark Ages
    # Middle Ages | Present
    # Future      | End of Time
    'All Eras (Vertical)': {
        'Prehistory': (0, 0),
        'Dark Ages': (1536, 0),
        'Middle Ages': (0, 1024),
        'Present': (1536, 1024),
        'Future': (0, 2048),
        'End Of Time': (1536, 2048),
    },
}


def get_all_locations():
    """Flattened list of all locations in locations.json"""
    locations_json = Path(Path(__file__).resolve().parent, "..", "locations/locations.json")
    data = json.loads(locations_json.read_text())

    # recursively generate flat list of all locations, recursing down childre
    def gen_locations(item):
        for child in item["children"]:
            if isinstance(child, t.Mapping) and "children" in child:
                for loc in gen_locations(child):
                    yield loc
            else:
                yield child

    return [loc for item in data for loc in gen_locations(item)]


ALL_LOCATIONS = get_all_locations()

# TESTS ######################################################################


@pytest.mark.parametrize("loc", ALL_LOCATIONS, ids=lambda loc: loc["name"])
def test_locations_coordinates(loc):
    """Test all locations era-relative coordinates have matching absolute coords."""
    # for loc in all_locations:
    map_locations = loc["map_locations"]
    abs_maps = MAP_OFFSETS.keys()
    spots = [spot for spot in map_locations if spot["map"] not in abs_maps]

    for abs_map, era in MAP_OFFSETS.items():
        abs_spots = [(spot["x"], spot["y"]) for spot in map_locations if spot["map"] == abs_map]

        # check that an absolute location exists
        assert abs_spots, f"No '{abs_map}' map location for {loc['name']}"

        # check that same number of absolute and era-relative locations
        assert len(spots) == len(
            abs_spots
        ), f"Inequal number of era-relative and '{abs_map}' map locations for {loc['name']}"

        # check each spot has corresponding absolute spot with correct offset
        for spot in spots:
            offset_x, offset_y = MAP_OFFSETS[abs_map][spot["map"]]
            expected = (spot["x"] + offset_x, spot["y"] + offset_y)
            assert expected in abs_spots, (
                f"Missing expected '{abs_map}' spot relative to '{spot['map']}' "
                f"for {loc['name']} with coords {expected}"
            )


scripts/init.lua
function getTrackerClient()
  if PopVersion then
    return {PopTracker=true}
  end
  return {EmoTracker=true}
end

TrackerClient = getTrackerClient()

if TrackerClient.PopTracker then
  print("Detected PopTracker: " .. PopVersion)
else
  print("Assuming EmoTracker")
end

ScriptHost:LoadScript("scripts/logic.lua")

--
-- Adds code watchers.
-- NOTE: This only works on PopTracker.
--
function addCodeWatchers()

  if not TrackerClient.PopTracker then
    print("Code watcher scripts not activated (incompatible tracker client).")
    return
  end

  print("Adding Code Watchers...")

  -- Toggles extra items on when "Shoe Extra Items" is toggled on
  function toggleExtraItems(_code)
    if hasFlagEnabled("ToggleExtraItems") then
      Tracker:AddLayouts("layouts/components/items_grid_extras.json")
    else
      Tracker:AddLayouts("layouts/components/items_grid.json")
    end
  end

  -- Toggles displaying "Flags" on Map Tracker (when off just in settings menu)
  function toggleHideFlags(_code)
    if hasFlagEnabled("ToggleHideFlags") then
      Tracker:AddLayouts("layouts/components/bottom_dock_hide_flags.json")
    else
      Tracker:AddLayouts("layouts/components/bottom_dock.json")
    end
  end

  ScriptHost:AddWatchForCode("ToggleExtraItems", "toggle_extra_items", toggleExtraItems)
  ScriptHost:AddWatchForCode("ToggleHideFlags", "toggle_hide_flags", toggleHideFlags)

end

--
-- Adds components used by tracker for items, bosses, flags, etc.
--
function addComponents()
  -- NOTE: Tracker will prefix components automatically with variant name
  -- and prefer those, but if they don't exist, will fall back to
  -- using paths relative to pack root directory

  print("Adding Components...")

  -- NOTE: Load order matters because Tracker cannot handle
  -- forward references consistently, so must load things first before can reference
  Tracker:AddLayouts("layouts/components/items_grid.json")
  Tracker:AddLayouts("layouts/components/bosses_grid.json")
  Tracker:AddLayouts("layouts/components/flags_grid.json")
  Tracker:AddLayouts("layouts/components/extra_flags_grid.json")
  Tracker:AddLayouts("layouts/components/options_grid.json")
  if Tracker.EmoTracker then
    Tracker:AddLayouts("layouts/components/settings_popup.emo.json")
  else
    Tracker:AddLayouts("layouts/components/settings_popup.json")
  end
  Tracker:AddLayouts("layouts/components/bottom_dock.json")

end

--
-- Adds main tracker and broadcast layouts based on mode and variant.
--
function addTrackerLayouts()

  print("Adding Layouts...")

  Tracker:AddLayouts("layouts/capture.json")

  if itemsOnlyTracking() then
    Tracker:AddLayouts("items_only/layouts/tracker.json")
  else
    Tracker:AddLayouts("layouts/tracker.json")
  end

  Tracker:AddLayouts("layouts/broadcast.json")

end

-- Configure tracker
print("Configuring Tracker...")
Tracker:AddItems("items/items.json")
Tracker:AddMaps("maps/maps.json")
Tracker:AddLocations("locations/locations.json")
addComponents()
addTrackerLayouts()
addCodeWatchers()

if _VERSION == "Lua 5.3" then
  print("Setting up Autotracking...")
  ScriptHost:LoadScript("scripts/autotracking.lua")
else
  print("Auto-tracker is unsupported by your tracker version")
end


tools/release.sh
#!/usr/bin/env bash

# This script is used for building a zip file suitable for PopTracker/Emotracker.
# 
# It builds a zip file which abides excluding files per .gitattributes (like git archive)
# but dereferences all symlinks (making copies of their linked files).

set -euo pipefail

CWD="$(builtin pwd)"
GIT_ROOT="$(builtin cd "$(dirname "${BASH_SOURCE[0]}")"/.. && builtin pwd)"
export CWD GIT_ROOT

function cleanup {
  ret=$?
  if [[ -z "${SKIP_CLEANUP:-}" ]]; then
    if [[ -n "${TEMP_DIR:-}" ]]; then 
      rm -rf "${TEMP_DIR}"
    fi
  else
    if [[ -n "${TEMP_DIR:-}" ]]; then 
      echo "Skipping cleanup of temporary files in ${TEMP_DIR}"
    fi
  fi
  exit $ret
}

trap cleanup EXIT

function preflight {
  # assure that tools used by this script are installed
  commands=(git tar zip unzip)
  for cmd in "${commands[@]}"; do
    if ! command -v "$cmd" > /dev/null; then
      >&2 echo -e "\e[31mThis script requires $cmd!\e[0m"
      exit 1
    fi
  done 

  if [[ -z "${RELEASE_ZIP}" ]]; then
    # if no specified name, generate a zip name
    if [[ -n "${CUSTOMIZE_SCRIPT}" ]]; then
      RELEASE_ZIP="Jets-of-Time-Tracker-custom.zip"
    else
      RELEASE_ZIP="Jets-of-Time-Tracker-$(git rev-parse --short HEAD).zip"
    fi
    export RELEASE_ZIP
  fi

  # check if release zip already exists and prompt to overwrite
  if [[ -e "${RELEASE_ZIP}" ]]; then
    if [[ -n "${YES}" ]]; then
      echo "Overwriting existing ${RELEASE_ZIP}"
    else
      read -p "Overwrite existing ${RELEASE_ZIP}? [y/N] " -n 1 -r
      if [[ ! "$REPLY" =~ ^[Yy]$ ]]; then exit 0; fi
      echo
    fi
  fi
}

function build_release {
  echo -e "\e[1;30mBuilding release:\e[0m ${RELEASE_ZIP}"

  TEMP_DIR="${TEMP_DIR:-$(mktemp -d)}"
  export TEMP_DIR
  echo "* Building in ${TEMP_DIR}"

  # create an archive just to get a list of files that abides .gitattributes
  echo "* Creating git archive"
  cd "$GIT_ROOT" || (>&2 echo -e "\e[31mCan't change to $GIT_ROOT directory!\e[0m" && exit 1)
  git archive HEAD --format=tar -o "${TEMP_DIR}/archive.tar"

  # use list of non-directory files from initial archive tarball to produce a new tar
  # with symlinks dereferenced into copies, so can with EmoTracker
  echo "* Creating dereferenced tarball"
  tar -tf "${TEMP_DIR}/archive.tar" | grep -v '/$' > "${TEMP_DIR}/files.txt"
  tar -chf "${TEMP_DIR}/release.tar.gz" -T "${TEMP_DIR}/files.txt"

  # extract the new tar and zip it up
  echo "* Building release zip"
  cd "${TEMP_DIR}"
  rm -rf zip
  mkdir zip
  cd zip
  tar -xf ../release.tar.gz

  # optional customize.sh if user wants to use it
  if [[ -n "${CUSTOMIZE_SCRIPT}" ]] && [[ -x "${CUSTOMIZE_SCRIPT}" ]]; then
    echo "* Applying customizations from ${CUSTOMIZE_SCRIPT}"
    cp "${CUSTOMIZE_SCRIPT}" "${TEMP_DIR}/customize.sh"
    pwd="$(builtin pwd)"
    cd "${TEMP_DIR}"
    ./customize.sh
    cd "$pwd"
  fi

  zip -q -r ../release.zip .
  cd "${CWD}"
  cp "${TEMP_DIR}/release.zip" "${RELEASE_ZIP}"

  # verify zip
  echo "* Verifying release zip"
  unzip -qt "${RELEASE_ZIP}" > /dev/null

  echo -e "\e[1;30mCreated release:\e[0m ${RELEASE_ZIP}"
}

function usage {
  echo "./tools/release.sh [-c] [-o] [-y]"
  echo
  echo "-c     run customization script as part of build"
  echo "-h     print this help"
  echo "-o     output zip filename to use"
  echo "-y     do not prompt (overwrites release zip)"
}

CUSTOMIZE_SCRIPT="${CUSTOMIZE_SCRIPT:-}"
RELEASE_ZIP="${RELEASE_ZIP:-}"
YES="${YES:-}"
while getopts 'c:ho:y' flag; do
  case "${flag}" in
    c)
      CUSTOMIZE_SCRIPT="${OPTARG}"
      if [[ -n "${CUSTOMIZE_SCRIPT}" ]] && [[ ! -x "${CUSTOMIZE_SCRIPT}" ]]; then
        >&2 echo -e "\e[31mCustomize script is not executable: ${CUSTOMIZE_SCRIPT}\e[0m"
        exit 1
      fi
      ;;
    h)
      usage
      exit 0
      ;;
    o)
      RELEASE_ZIP="${OPTARG}"
      ;;
    y)
      YES=1
      ;;
    *)
      usage
      exit 1
  esac
done
export CUSTOMIZE_SCRIPT RELEASE_ZIP YES

preflight
build_release


tools/customize.example.sh
#!/usr/bin/env bash

set -euo pipefail

# This script is intended to be used to provide user-specific customizations
# to pack files before they are zipped up by tools/release.sh.
#
# If this script was copied to "customize.sh", a custom pack could be built
# via:
#
#     ./tools/release.sh -c customize.sh -o ctjot-custom.zip
#
# It is run within the $TEMP_DIR directory, immediately before the final
# "Building release zip" phase of tools/release.sh. It has access to all the
# pack files in the "zip" directory inside of $TEMP_DIR.
#
# The following environment variables are guarnteed to be available:
#
# TEMP_DIR : location of temporary build files (where this script is running)
# GIT_ROOT : the location of git files for this repo (outside of TEMP_DIR)


function regen_images {
  # Regenerates images using tools/generate-flag-images.sh with custom settings
  echo " -> Regenerating flag images"
  FONT="Cantarell-Extra-Bold"
  OUTDIR="${TEMP_DIR}/zip/images/flags"
  export FONT FLAG_OFF FLAG_ON OUTDIR
  "${GIT_ROOT}/tools/generate-flag-images.sh" -fy >/dev/null
}

# update manifest.json for custom version
if command -v jq; then
  contents="$(jq '.name |= "Jets of Time Tracker (Customized)"' zip/manifest.json)"
  echo "$contents" > "zip/manifest.json"
fi

if command -v magick; then
  regen_images
fi


tools/normalize-locations.py
#!/usr/bin/env python
import argparse
import copy
import json
import sys
import typing as t

from pathlib import Path

MAP_OFFSETS: t.Dict[str, t.Tuple[int, int]] = {
    # Prehistory | Middle Ages | Future
    # Dark Ages  | Present     | End of Time
    'All Eras': {
        'Prehistory': (0, 0),
        'Dark Ages': (0, 1024),
        'Middle Ages': (1536, 0),
        'Present': (1536, 1024),
        'Future': (3072, 0),
        'End Of Time': (3072, 1024),
    },
}


class CLI:
    def __init__(self):
        self.args = self.get_parser().parse_args()

    @staticmethod
    def get_parser():
        parser = argparse.ArgumentParser(
            prog='normalize-locations.py',
            description='Normalize absolute map locations based on era in locations.json',
        )
        parser.add_argument('locations_file')
        parser.add_argument('-i', '--in-place', help='modify file in-place (overwrites file)', action='store_true')
        parser.add_argument('-k', '--check', help='return non-zero exit if needs updates', action='store_true')
        parser.add_argument('-x', '--explain', help='explain updates to be made', action='store_true')
        return parser

    def gen_coordinates(self, location: t.Dict[str, t.Any]) -> t.Dict[str, t.Any]:
        '''Yield era-relative and absolute ('All Eras') locations.'''
        map_locations = location['map_locations']
        eras = [era for era in map_locations if era['map'] not in ['All Eras', 'All Eras (Vertical)']]

        for era in eras:
            yield era

            for absolute_map, era_offsets in MAP_OFFSETS.items():
                offset_x, offset_y = era_offsets[era['map']]
                absolute = {'map': absolute_map, 'x': era['x'] + offset_x, 'y': era['y'] + offset_y}

                if self.args.explain and absolute not in map_locations:
                    print(f"Updating '{location['name']}' -> {absolute}")

                yield absolute

    def update_coordinates(self, location):
        if 'children' in location:
            for child in location['children']:
                self.update_coordinates(child)
        else:
            location.update({'map_locations': [loc for loc in self.gen_coordinates(location)]})

    def update_locations(self, data) -> t.List[t.Dict[str, t.Any]]:
        updated = copy.deepcopy(data)
        for item in updated:
            self.update_coordinates(item)
        return updated


def main(cli: CLI) -> int:
    data = json.loads(Path(cli.args.locations_file).read_text())
    updated = [item for item in cli.update_locations(data)]
    output = json.dumps(updated, indent=2) + '\n'

    if cli.args.in_place:
        if data == updated:
            print(f"No modifications found, not overwriting {cli.args.locations_file}")
        else:
            Path(cli.args.locations_file).write_text(output)
            print(f"Modified {cli.args.locations_file}")
    elif not cli.args.explain:
        print(output)

    if cli.args.check and data != updated:
        print(f"File {cli.args.locations_file} needs updates.")
        return 1
    return 0


if __name__ == '__main__':
    cli = CLI()
    sys.exit(main(cli))
